### PR TITLE
77 support direct download

### LINKED ISSUE #77: Support direct download
Currently, to download a programme you need to refresh, then search through the list of programmes.

@revad has requested direct download, so that if you already know the pid, or the URL on the BBC site, you can enter this and request to download the media directly. This would be a nice addition.

This could be added in as a separate dialog, available through the pulley menu on the queue screen. "Direct download", or something along those lines.

### PR DESCRIPTION
**Pull request by llewelld**
Closes #77.

### CODE REVIEW

File: src/queue.cpp [hunk 1]
```diff
@@ -303,3 +303,14 @@ quint32 Queue::getMediaPosition(QString progid) {
 
     return position;
 }
+
+QString Queue::extractPid(QString progid) {
+    QString result = "";
+
+    QRegExp search("^(http.*/)?(\\w{8})$");
```
Comment (dinkypumpkin) [inline src/queue.cpp]: I think this will fail for World Service programmes that have 15-char PIDS.

Comment (llewelld) [reply src/queue.cpp]: Yes, you're absolutely correct. I just based this on my own limited experience and assumed exactly 8-character PIDs. Is it safe to assume either 8 or 15 characters? All of the PIDs I've seen so far start with `b` or `w`; is it safe to assume anything like this?

Comment (dinkypumpkin) [reply src/queue.cpp]: Only 8 or 15 in use afaik, but wasn't always the case. I suggest you just use 8 as minimum. Many PIDs begin with `p` as well, but you don't need to use that. Instead, take advantage of the fact that PIDs only contain digits and consonants to make a more specific regex pattern. And remember that iPlayer TV episode URLs don't end  with a PID.

Comment (llewelld) [reply src/queue.cpp]: Great, thank you, this is really useful info. It hadn't occurred that the vowels were all missing (I thought there was something peculiar going on but having ruled out hex, I couldn't put my finger on it!).

**Pull request merged**